accepted/install-locations.md

```diff
@@ -0,0 +1,227 @@
+  Note that this registry key is "redirected" that means that 32-bit processes see different copy of the key then 64bit processes. So it's important that both installers and the host access only the 32-bit view of the registry.
+* Linux and macOS - the installer will store the install location in a text file
+  * `/etc/dotnet/install_location` - the file will contain a single line which is the path.
+
+The muxer (`dotnet.exe`) is registered on the system-wide `PATH` environment variable.
+
+### Per-user registered install location :new:
+This will allow installs to per-user directories (no need for admin/root access). It's to be decided if we will actually create an installer which does this. Once installed the per-user install should take precedence over the global one, but otherwise should be available to all processes running under the respective user.
+
+To achieve this the install location must be recorded in some per-user configuration in a fixed location which is well-known.
+
+* Windows - the installer will store the install location in registry:
+  * `HKCU\SOFTWARE\dotnet\Setup\InstalledVersions\<arch>\InstallLocation` - where `<arch>` is one of the `x86`, `x64`, `arm` or `arm64`.
+  Note that this key is "shared" meaning both 32-bit and 64-bit processes see the same values.
+* Linux and macOS - the installer will store the install location in a text file
+  * `$HOME/.dotnet/install_location` - the file will contain a single line which is the path.
+
+The muxer (`dotnet`) is registered on the per-user `PATH` environment variable and in such a way that it wins over any globally installed one.
+
+### Private install location
+This allows users to install private copy of .NET Core into any directory. This copy should not be used by anything unless explicitly asked for.
+
+There is not installer for this scenario, these installs are produced by simply copying the right directory structure into a custom location.
+
+*This mechanism already exists in .NET Core 2.2.*
+
+If the muxer (`dotnet`) is used then it must be invoked directly from the private install location (it's not registered on `PATH` in any way).  
+If other hosts are used then the private install location must be specified in environment variable:
+* `DOTNET_ROOT`
+* `DOTNET_ROOT(x86)` for 32-bit process running on 64-bit Windows.
+
+
+## Host behavior for 3.0
+
+### `hostfxr` search
+`dotnet` only looks right next to itself.
+
+All other hosts will search for the first location which exists in this order:
+* the directory where the host resides (this is important mostly for `apphost` as self-contained apps will have the `hostfxr` in the same directory).
+* `DOTNET_ROOT` (or `DOTNET_ROOT(x86)`) environment variable, in the `host/fxr/<version>` subdirectory.
+* :new: per-user registered install location, in the `/host/fxr/<version>` subdirectory
+  * Windows - `HKCU\SOFTWARE\dotnet\Setup\InstalledVersions\<arch>\InstallLocation`
```

Comment:
I would like to hear more about the motivation behind this design.  What if this user writable key points to a location in my temp directory with a fake hostfxr and later used by an elevated process?

Reply:
Good point - didn't realize it has security implications - as mentioned above, I'll remove this.

Reply:
I think that the problem with multi-level lookup is that it mixes local unzipped installs with global ones. If it stopped doing that, it would be fine with me (on both Windows and non-Windows).

Would something like this work (on both Windows and non-Windows)?

- For `dotnet` run from official global install, stitch together all global locations. (We would need a bread crumb that indentifies official global install.)
- For `dotnet` run from local unzipped install, look in the directory where `dotnet` resides or `DOTNET_ROOT` points to only.